### ISSUE TITLE
CanvasLMS RCE CORS Issue

### ISSUE THREAD

**Issue by OpenTechNW** [bitnami/vms#964]
### Platform

AWS

### bndiagnostic ID [know more about bndiagnostic ID](https://docs.bitnami.com/general/how-to/understand-bndiagnostic/)

8ec5138b-0164-25a5-53b0-061c2150b45c

### bndiagnostic output

===== Begin of bndiagnostic tool output =====

 
    ✓ Resources: No issues found
    ? Connectivity: Found possible issues
    ✓ Processes: No issues found
    ? Apache: Found possible issues
 

 
[Connectivity]

Server ports 22, 80 and/or 443 are not publicly accessible. Please check the 
following guide to open server ports for remote access:

https://docs.bitnami.com/general/faq/administration/use-firewall/

[Apache]

Found recent error or warning messages in the Apache error log.
```
[Mon Jun 05 16:28:23.720409 2023] [proxy:error] [pid 795450:tid 140230983890688] 
Press [Enter] to continue:
[client **ip_address**:3965] AH00898: DNS lookup failure for: 
127.0.0.1:3000setup.cgi returned by /setup.cgi

[Mon Jun 05 19:36:12.859211 2023] [proxy:error] [pid 795450:tid 140231076210432] 
[client **ip_address**:53211] AH00898: DNS lookup failure for: 
127.0.0.1:3000setup.cgi returned by /setup.cgi

[Mon Jun 05 20:25:46.459588 2023] [proxy:error] [pid 820755:tid 140231092995840] 
[client **ip_address**:41226] AH00898: DNS lookup failure for: 
127.0.0.1:3000favicon.ico returned by /favicon.ico, referer: 
https://rce.psychcore-learning.net/

```
Please check the following guide to troubleshoot server issues:
 
https://docs.bitnami.com/general/apps/wordpress/troubleshooting/debug-erro
rs-apache/
 

===== End of bndiagnostic tool output =====

### bndiagnostic was not useful. Could you please tell us why?

partly

### Describe your issue as much as you can

among other issues, i cannot upload files or import courses. 
i get the following CORs error

Access to fetch at 'https://rce.mydomain.net/api/session' from origin 'https://canvas.mydomain.net' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

have tried adding 'Access-Control-Allow-Origin' to apache configs and .htaccess and still get the error.

RCE seems to be setup properly, as i get the hello from rce message when i go to https://rce.mydomain.net/

what do i need to add where to make this work?

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Thanks for using Bitnami. I checked your bnsupport information and I see you have a line for `Access-Control-Allow-Origin "*"`, although the line is not fully correct. The line you need to add is the one below

```
Header set Access-Control-Allow-Origin "*"
```

You will have to include it in the `<Directory>` section in the `apache/conf/vhosts/canvaslms-vhost.conf` and `apache/conf/vhosts/canvaslms-https-vhost.conf` files. After that, remember to restart Apache service for changes to take effect

```
sudo /opt/bitnami/ctlscript.sh restart apache
```

Hope it helps!

**Comment by OpenTechNW** [comment]
Thanks, that is set but it still doesn't work
I also set it in the canvas-rce-api-vhost.conf and canvas-rce-api-https-vhost.conf file and still get the same issue

**Comment by OpenTechNW** [comment]
also when i go to https://rce.mydomain.net/api/session

Proxy Error
The proxy server received an invalid response from an upstream server.
The proxy server could not handle the request

Reason: DNS lookup failure for: 127.0.0.1:3000api

**Comment by OpenTechNW** [comment]
Any ideas? the best i can find it that it has something to do with the preflight policies but cant get anything to take and cant find where that is set in the canvas app

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Thanks for your messages. Can you run the bndiagnostic tool again and share the new code with us so we can check the updated configuration and error logs?

**Comment by OpenTechNW** [comment]
702111f4-cb21-6a26-7a75-8bc3d0663f14

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Thanks for the new code. I still see the CORS configuration in the `canvaslms-vhost.conf` and `canvaslms-https-vhost.conf` files outside the section defined between `<Directory "/opt/bitnami/canvaslms/public">` and `</Directory>`. Can you move the configuration inside the section, restart Apache for changes to take effect, and then give it another try?

**Comment by OpenTechNW** [comment]
I moved the config inside the <Directory> section and it still does not work

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Did you restart Apache after that? It is weird that the configuration is in the correct place but the application is not working. I don't see any error message related to this in the latest bnsupport bundle you shared. 

However, I see one error that caught my attention

```
[Mon Jun 12 17:00:15.215629 2023] [proxy:error] [pid 234353:tid 140051392046848] [client CLIENT-IP:24273] AH00898: DNS lookup failure for: 127.0.0.1:3000api returned by /api/session, referer: https://stage.YOUR-DOMAIN.net/courses/23/pages/start-page/edit
```

I don't know how that `127.0.0.1:3000api` string is generated, but it looks wrong to me. Can you check it in the CanvasLMS project? Apart from that, I think that asking about the CORS settings and how to better debug it would be of help.

https://community.canvaslms.com/t5/Canvas-Question-Forum/bd-p/questions

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Can you run the bndiagnostic tool again so we can take a look to the updated Apache logs?

**Comment by OpenTechNW** [comment]
new code: e8c6eeaf-88cd-05c6-d491-46cf9db1b38e

**Comment by gongomgra** [comment]
Hi @OpenTechNW,

Thanks for the update. You Apache config looks fine to me. I also see the next error in the Apache logs, although I don't know if it is related to the CanvasLMS RCE CORS issue you are facing or not

```
[ N 2023-06-19 19:58:49.9699 588705/T3 age/Cor/CoreMain.cpp:1146 ]: Checking whether to disconnect long-running connections for process 589994, application /opt/bitnami/canvaslms (production)
[ E 2023-06-19 21:45:11.3616 588718/T0 apa/Hooks.cpp:750 ]: Unexpected error in mod_passenger: An error occurred while receiving HTTP upload data: Partial results are valid but processing is incomplete (70008)
  Backtrace:
     in 'void Passenger::Apache2Module::Hooks::sendRequestBody(const Passenger::FileDescriptor&, request_rec*, bool)' (Hooks.cpp:1230)
     in 'int Passenger::Apache2Module::Hooks::handleRequest(request_rec*)' (Hooks.cpp:622)
```

Let's wait for the official developers. Sorry for the inconveniences.

**Comment by eharvey71** [comment]
I ran into this issue and there is still an additional problem that needs to be resolved on the Bitnami image for Canvas.

I just deployed from the latest image into AWS last week (10/03/23). The CORS-related fix above is necessary and worked for me but theres's still a problem with the encryption secret and signing_secret having been recently moved from dynamic_settings.yml to vault_contents.yml. 

A notice from the Canvas dev team was posted here: https://github.com/instructure/canvas-lms/blob/master/packages/canvas-rce/DEVELOPMENT.md#canvas-rce-api

... but that post seems to still give an incomplete rundown of what the final vault_contents.yml should have in it. Here's what fixed this for me when moving over the secrets from one file to the other:

**dynamic_settings.yml**
production:
  config:
    canvas:
      rich-content-service:
        app-host: "rce.mydomain.com"

**vault_contents.yml**
production:
  'app-canvas/data/secrets':
    data:
      canvas_security:
        encryption_secret: "astringthatisactually32byteslong"
        signing_secret: "astringthatisactually32byteslong"

I really hope this helps others. With issues concerning LTI Configuration, SIS Import, and outbound email -- this has been a learning experience!
